Array.shuffle: explicitly validate the result of [rand].

fpottier opened the pull request:
This yields more readable code (the previous code used [get] instead of [unsafe_get] in a way that was quite cryptic), a more precise runtime check (in the previous code, a violation of [rand]'s contract could go unnoticed), and a more understandable exception (namely, [Invalid_argument "Array.shuffle"] instead of
[Invalid_argument "array index out of bounds"]).

fpottier committed:
Array.shuffle: explicitly validate the result of [rand].

This yields more readable code (the previous code used [get] instead
of [unsafe_get] in a way that was quite cryptic), a more precise
runtime check (in the previous code, a violation of [rand]'s contract
could go unnoticed), and a more understandable exception (namely,
[Invalid_argument "Array.shuffle"] instead of
[Invalid_argument "array index out of bounds"]).

dbuenzli commented on stdlib/array.ml:
```diff
@@ -442,8 +442,9 @@ let fast_sort = stable_sort
 let shuffle ~rand a = (* Fisher-Yates *)
   for i = length a - 1 downto 1 do
     let j = rand (i + 1) in
+    if not (0 <= j && j <= i) then invalid_arg "Array.shuffle";
```
Good idea, but perhaps we could try to slowly move the Stdlib towards less cryptic error messages that require less trips to printf debugging. 

Here are gradual improvements. 
```ocaml
invalid_arg "Array.shuffle: rand"
invalid_arg "Array.shuffle: rand returned an unexpected integer"
invalid_arg ("Array.shuffle: rand " ^ (string_of_int (i + 1)) ^ " returned " ^
  (string_of_int j) ^ ", out of expected [0;" ^ (string_of_int i) "] range")
```

Personally I would really like to have the last one which has all the information one would ask for in a debug session. It is equivalent to the format string: 
```
"Array.shuffle: rand %d returned %d, out of expected [0;%d] range"
```
but I suspect we don't want (or can't) to get a `Printf` dependency on `Array`. If that's actually ok defining `invalid_argf` as `Printf.ksprintf invalid_arg` is always nice for formatted raises. 

Having a separate function like `err_rand` for formatting the message or the raise allows to keep the function clean from the nice but long morasses of text (and allows error message reuse of course).

fpottier replied:
Thanks @gasche and @dbuenzli for your suggestions. I have improved the error message (without depending on `Printf`) and have placed it in a separate function.

Octachron replied:
Rather than using `Printf` or `^`, I would propose to use  `String.concat`:
```ocaml
    String.concat "" [
      "Array.shuffle: rand("; string_of_int (i + 1); ")"; " returned "; string_of_int j;
      ", outside of the expected range [0; "; string_of_int i; "]"
   ]
 ```
 mostly because a punctuation like `;` feels less intrusive than `^`.
(hm, that's one case where I would have used `[: ... :]` if available)

ferminr commented on stdlib/array.ml:
```diff
@@ -442,8 +442,9 @@ let fast_sort = stable_sort
 let shuffle ~rand a = (* Fisher-Yates *)
   for i = length a - 1 downto 1 do
     let j = rand (i + 1) in
+    if not (0 <= j && j <= i) then invalid_arg "Array.shuffle";
```
How about this version instead?
`if j < 0 || j > i then invalid_arg "Array.shuffle";`

gasche replied:
I would want a better error message than this, for users to know what is going on. For example `"Array.shuffle: `rand` returned an integer out of range"`. Or one could even be more clear:

```ocaml
Printf.ksprintf invalid_arg
  "Array.shuffle: `rand` returned an integer %d \
   out of the expected range [0; %d]"
  j i
```

fpottier committed:
Array.shuffle: improve the error message that is produced when [rand] violates its contract.

dbuenzli commented on stdlib/array.ml:
```diff
@@ -439,10 +439,18 @@ let stable_sort cmp a =
 
+let shuffle_contract_violation i j =
+  let msg =
+    "Array.shuffle: rand(" ^ string_of_int (i + 1) ^ ")" ^
```
That's not an idiomatic OCaml function call which is going to look weird when you get the error. I mean in `man` of my system `rand(3)` is actually a thing :-) If you feel parens are needed then I rather have `(rand %d)` and or use ocamldoc syntax `[rand %d]` but then we should establish a pattern here.

dbuenzli replied:
Yet another way would be to use the compiler's convention for quoting code which I believe is straight quotes nowadays.

dbuenzli commented on stdlib/array.ml:
```diff
@@ -439,10 +439,18 @@ let stable_sort cmp a =
+  let msg =
+    "Array.shuffle: rand(" ^ string_of_int (i + 1) ^ ")" ^
+    " returned " ^ string_of_int j ^
+    ", outside of the expected range [0; " ^ string_of_int i ^ "]"
```
note that using the more brief `out of expected range` makes it slightly more probable that the error message fits on a line (assuming 80 columns) and is thus easier to read.

hannesm commented:
I have a question about how defensive should the OCaml stdlib be with respect to existing functions in other modules? So, where should checking end, where should it begin? Should a "String.make" be validated to be of the right size (and contain the right data)? I'm just curious where to start and where to stop.

gasche commented:
Here the `rand` function is a parameter provided by the user, not the call to another function in the standard library, and we need to check in any case that the result is a valid index in the array.

hannesm commented:
@gasche ah, thanks. I didn't read the code carefully enough.

fpottier committed:
[shuffle_contract_violation]: change wording of error message.

Use straight quotes as delimiters of call, as in 'rand 3',
instead of parentheses as delimiters of argument, as in rand(3).

Shorten the message by writing "out of expected range".

fpottier committed:
[shuffle_contract_violation]: use [String.concat].

Also, abbreviate [string_of_int], for readability.

fpottier commented:
Thanks all. I have taken your comments into account.

dbuenzli reviewed (approved):
Thank you! That's what should have been done in the first place (by me).

gasche commented:
@fpottier could you add a Changes entry?

Octachron commented:
The dependencies also needs to be updated with `make alldepend` to account for the `Array` -> `String` dependency.

fpottier committed:
Changes entry.

fpottier committed:
Update stdlib/.depend with a dependency of Array on String.

fpottier commented:
OK, done. I have updated just `stdlib/.depend`.

`make alldepend` also caused changes in `.depend`, but they seem unrelated to my changes.
